QemuPkg/Plugins/VirtualDriveManager/VirtualDriveManager.py

```diff
@@ -0,0 +1,410 @@
+        filename = Path(filepath).name
+
+        cmd = "FileInsert"
+        args = f"{filepath} {filename} {str(self.drive_path)}"
+        result = RunCmd(cmd, args)
+        if result != 0:
+            e = f"[{cmd} {args}] Result: {result}"
+            logger.error(f"Failed to insert {filepath} into drive.")
+            logger.error(e)
+            raise RuntimeError(e)
+
+    def get_file(self, virtual_path: PathLike, local_path: PathLike):
+        """Gets a file from the virtual drive.
+        
+        Args:
+            virtual_path (PathLike): The path to the file on the virtual drive
+            local_path (PathLike): The path to save the file to
+        
+        Raises:
+            (RuntimeError): Failed to get the filepath
+        """
+        cmd = "FileExtract"
+        args = f"{virtual_path} {local_path} {self.drive_path}"
+        result = RunCmd(cmd, args)
+        if result != 0:
+            e = f"[{cmd} {args}] Result: {result}"
+            logging.error(f"Failed to get {virtual_path} from drive.")
+            logging.error(e)
+            raise RuntimeError(e)
+
+    def get_file_contents(self, virtual_path: PathLike, local_path: PathLike = None):
+        """Gets a contents from a file from the virtual drive. Optionally save the file too.
+        
+        Args:
+            virtual_path (PathLike): The path to the file on the virtual drive
+            local_path (PathLike): The path to save the file to
+        
+        Raises:
+            (RuntimeError): Failed to get the filepath
+        """
+        if local_path is None:
+            local_path = tempfile.mktemp()
+
+        self.get_file(virtual_path, local_path)
+
+        with open(local_path, "rb") as f:
+            return f.read()
+
+
+class VirtualDriveManager(IUefiHelperPlugin):
+    def RegisterHelpers(self, obj):
+        fp = str(Path(__file__).absolute())
+        obj.Register("get_virtual_drive", VirtualDriveManager.get_virtual_drive, fp)
+        obj.Register("add_tests", VirtualDriveManager.add_tests, fp)
+        obj.Register("report_results", VirtualDriveManager.report_results, fp)
+        return 0
+    
+    @staticmethod
+    def get_virtual_drive(path: PathLike):
+        if os.name == 'nt':
+            return WindowsVirtualDrive(path)
+        return LinuxVirtualDrive(path)
+    
+    @staticmethod
+    def add_tests(drive: VirtualDrive, test_list: list[str], auto_run = True, auto_shutdown = True):
+        """Adds tests to the virtual drive and optionally adds them to the startup script.
+        
+        Args:
+            drive (VirtualDrive): The virtual drive to add the tests to.
+            auto_run (Boolean): Whether or not to run tests automatically.
+            auto_shutdown (Boolean): Whether or not to shutdown after tests have completed.
+        """
+        drive.add_files(test_list)
+
+        if auto_run:
+            tests = []
+            # Execute all tests
+            for test in test_list:

```

Comment:
what is going on here?  This looks like some method to get around the "cache.dat" stuff.

Reply:
ok.  after reading the PR history i now understand this.  Might make sense to add a comment to explain the complexity in the startup nsh and the solution to optimize the tests running

Reply:
There are two parts to this:

1. This is a solution to reduce test times by preventing the efi image from loading if the test has already run. Before this change, any test that requires a restart would force the system to reload all previous test images and execute them. While the tests are skipped due to the presence of a _Cache.dat file, it still takes time to load the image.

1. We want to remove the *_Cache.dat file at the end of the run so that future runs (where EMPTY_DRIVE=FALSE) will actually run the test instead of loading the test, seeing that it has already run, and skipping.